pkg/certificate/providers/mrc.go

```diff
@@ -0,0 +1,118 @@
+	MRCProviderGenerator
+}
+
+// List returns the MRCs stored in the informerCollection's store
+func (m *MRCComposer) List() ([]*v1alpha2.MeshRootCertificate, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+	// informers return slice of pointers so we'll convert them to value types before returning
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	var mrcs []*v1alpha2.MeshRootCertificate
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcs = append(mrcs, mrc)
+	}
+
+	return mrcs, nil
+}
+
+// Watch returns a channel that receives events whenever MRCs are added, updated, and deleted
+// from the informerCollection's MRC store. Channels returned from multiple invocations of
+// Watch() are unique and have no coordination with each other. Events are guaranteed
+// to be ordered for any particular resources, but NOT across different resources.
+// Watch() should only send new events as of the time the Watch call returns
+func (m *MRCComposer) Watch(ctx context.Context) (<-chan certificate.MRCEvent, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+
+	// To make sure we only send new events, we cache the object versions
+	// before the eventhandler is added
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	mrcObjectVersionCache := map[string]int64{}
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcObjectVersionCache[mrcKeyFunc(*mrc)] = mrc.GetGeneration()
+	}
+
+	eventChan := make(chan certificate.MRCEvent)
+	m.informerCollection.AddEventHandler(informers.InformerKeyMeshRootCertificate, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			mrc := obj.(*v1alpha2.MeshRootCertificate)
+			if !checkGeneration(mrcObjectVersionCache, *mrc) {
+				return
+			}
+			eventChan <- certificate.MRCEvent{
+				Type: certificate.MRCEventAdded,
+				MRC:  mrc,
+			}
+		},
+		// We don't really care about the previous version
+		// since the "state machine" of the MRC is well defined
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			mrc := newObj.(*v1alpha2.MeshRootCertificate)
+			if !checkGeneration(mrcObjectVersionCache, *mrc) {
+				return
+			}
+			eventChan <- certificate.MRCEvent{
+				Type: certificate.MRCEventUpdated,
+				MRC:  mrc,
+			}
+		},
+		// We don't care about deletes because the only delates that should
+		// happen come from the control plane cleaning up an old MRC. Our
+		// ValdatingWebhookConfiguration should prevent deletes from users
+		DeleteFunc: func(obj interface{}) {},
+	})
+
+	return eventChan, nil
+}
+
+func checkGeneration(mrcObjectVersionCache map[string]int64, mrc v1alpha2.MeshRootCertificate) bool {
```

Comment:
is this checking what we discussed, where the informers guarantee we don't need to do this?

Reply:
From the informer docs re: event handlers:
```A client added after `Run()` starts gets a startup batch of notifications of additions of the objects existing in the cache at the time that client was added; also, for every update to the SharedInformer's local cache after that client was added, eventually either the SharedInformer is stopped or that client is notified of that update.  Client notifications happen after the corresponding cache update and, in the case of a SharedIndexInformer, after the corresponding index updates. ```

This function makes sure the `certificate.Manager` doesn't try to recompute issuer state from the very beginning. Admittedly, the drift between when the MRC informer starts and when the Manager starts to watch is likely minimal, but this logic also prevents `Watch` from duplicating results retrieved from the manger's initial `List` call

Reply:
since watch will return the entire set, should we just remove the list call?

would that fix the issues you're describing? if not can you elaborate on the state drift you're referring to

Reply:
That gets into my [conversation with John](https://github.com/openservicemesh/osm/pull/4816#discussion_r898464225). In order to return a ready-to-use `Manager`, my approach was essentially to initialize the two issuers based on the MRC state at one point of time (i.e. a stable state). From there, the manager can receive future updates via the watch. This function ensures that the events sent to the `Manager` are in fact "future updates" and not old events

Reply:
i think the thing is that our state machine can be really dumb. In that it can see an active MRC, and know that it is the only active one, and act accordingly.

I'd advocate for simplicity vs flexibility here

Reply:
But then what happens on a control plane restart during an MRC rotation? There will be 2 MRCs, maybe one validating and one issuing. I can factor out the base case (`len(mrcs) == 1`)  if that's what you're saying, but I feel like allowing users of `MRCBroker` implementations to assume `Watch` events are all relevant is worth some of the complexity. FWIW, this is the part of this PR I'm least comfortable with, so if we can address some of the scenarios I've described, I'll be the first one to rip this out 😄